File: app/src/main/java/org/dhis2/data/forms/dataentry/fields/FieldViewModelFactory.java
package org.dhis2.data.forms.dataentry.fields;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.dhis2.form.model.FieldUiModel;
import org.dhis2.form.model.LegendValue;
import org.dhis2.form.model.RowAction;
import org.dhis2.form.ui.style.BasicFormUiModelStyle;
import org.hisp.dhis.android.core.common.FeatureType;
import org.hisp.dhis.android.core.common.ObjectStyle;
import org.hisp.dhis.android.core.common.ValueType;
import org.hisp.dhis.android.core.common.ValueTypeDeviceRendering;
import org.hisp.dhis.android.core.option.Option;
import org.hisp.dhis.android.core.program.ProgramStageSectionRenderingType;
import org.hisp.dhis.android.core.program.ProgramTrackedEntityAttribute;
import org.hisp.dhis.android.core.trackedentity.TrackedEntityAttribute;

import java.util.List;

import autovalue.shaded.org.checkerframework$.checker.nullness.qual.$NonNull;
import io.reactivex.Flowable;
import io.reactivex.processors.FlowableProcessor;

public interface FieldViewModelFactory {

    @NonNull
    FieldUiModel create(@NonNull String id,
                        @NonNull String label,
                        @NonNull ValueType valueType,
                        @NonNull Boolean mandatory,
                        @Nullable String optionSet,
                        @Nullable String value,
                        @Nullable String programStageSection,
                        @Nullable Boolean AllowFutureDate,
                        @NonNull Boolean editable,
                        @Nullable ProgramStageSectionRenderingType renderingType,
                        @Nullable String description,
                        @Nullable ValueTypeDeviceRendering fieldRendering,
                        @Nullable Integer optionCount,
                        @NonNull ObjectStyle objectStyle,
                        @Nullable String fieldMask,
                        @Nullable LegendValue legendValue,
                        List<Option> options,
                        @Nullable FeatureType featureType
    );

    @NonNull
    FieldUiModel createForAttribute(@$NonNull TrackedEntityAttribute trackedEntityAttribute,
                                    @Nullable ProgramTrackedEntityAttribute programTrackedEntityAttribute,
                                    @Nullable String value,
                                    boolean editable,
                                    List<Option> options);

    @NonNull
    FieldUiModel createSingleSection(String singleSectionName);

    @NonNull
    FieldUiModel createSection(String sectionUid, String sectionName, String description,
                               boolean isOpen, int totalFields, int completedFields, String rendering);

    @NonNull
    FieldUiModel createClosingSection();

    @NonNull
    Flowable<String> sectionProcessor();

    @NonNull
    FlowableProcessor<RowAction> fieldProcessor();
}

File: form/src/main/java/org/dhis2/form/ui/FieldViewModelFactory.kt
package org.dhis2.form.ui

import org.dhis2.commons.intents.CustomIntentAction
import org.dhis2.commons.orgunitselector.OrgUnitSelectorScope
import org.dhis2.form.model.EventCategory
import org.dhis2.form.model.FieldUiModel
import org.dhis2.form.model.OptionSetConfiguration
import org.dhis2.form.model.PeriodSelector
import org.hisp.dhis.android.core.common.FeatureType
import org.hisp.dhis.android.core.common.ObjectStyle
import org.hisp.dhis.android.core.common.ValueType
import org.hisp.dhis.android.core.common.ValueTypeDeviceRendering
import org.hisp.dhis.android.core.program.SectionRenderingType
import org.hisp.dhis.mobile.ui.designsystem.component.SelectableDates

interface FieldViewModelFactory {
    fun create(
        id: String,
        label: String,
        valueType: ValueType,
        mandatory: Boolean,
        optionSet: String? = null,
        value: String? = null,
        programStageSection: String? = null,
        allowFutureDates: Boolean? = null,
        editable: Boolean,
        renderingType: SectionRenderingType? = null,
        description: String?,
        fieldRendering: ValueTypeDeviceRendering? = null,
        objectStyle: ObjectStyle = ObjectStyle.builder().build(),
        fieldMask: String? = null,
        optionSetConfiguration: OptionSetConfiguration? = null,
        featureType: FeatureType? = null,
        autoCompleteList: List<String>? = null,
        orgUnitSelectorScope: OrgUnitSelectorScope? = null,
        selectableDates: SelectableDates? = null,
        eventCategories: List<EventCategory>? = null,
        periodSelector: PeriodSelector? = null,
        customIntentAction: CustomIntentAction? = null
    ): FieldUiModel

    fun createSingleSection(singleSectionName: String): FieldUiModel

    fun createSection(
        sectionUid: String,
        sectionName: String?,
        description: String?,
        isOpen: Boolean,
        totalFields: Int,
        completedFields: Int,
        rendering: String?,
    ): FieldUiModel

    fun createClosingSection(): FieldUiModel
}
